### PR TITLE
Fix ddpg pytorch

### LINKED ISSUE #9667: [RLlib] Bugs in PyTorch version of DDPG
Hello,

I spotted some bugs in the implementation of DDPG in PyTorch.
* The gradient clipping is not implemented correctly, it uses the 'grad_norm_clipping' parameter instead of 'grad_clip' and the function `ray.rllib.utils.torch_ops.minimize_and_clip` which is not correct. All the others algorithms seems to use `ray.rllib.agents.a3c.a3c_torch_policy.apply_grad_clipping`. I propose to replace minimize_and_clip by the A3C function in torch_ops.

* In the PyTorch model of DDPG the parameters to bound the action space (range and minimum) are not tracked by PyTorch as they are not registered as parameters. This means that they are not converted to cuda tensors resulting in an error.

* The target model is placed on the gpu even if ray was not configure to use the gpu.

I will make a PR with everything. But I don't know if I should replace minimize_and_clip. 

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x]  I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

### PR DESCRIPTION
**Pull request by raphaelavalos**
Fix #9667 

@sven1977 here is the PR :)

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/agents/ddpg/ddpg_torch_model.py [hunk 1]
```diff
@@ -88,9 +88,16 @@ def __init__(self,
         # emulate behaviour of tanh activation used in DDPG and TD3 papers.
         # After sigmoid squashing, re-scale to env action space bounds.
         class _Lambda(nn.Module):
+            def __init__(self_):
+                super(_Lambda, self_).__init__()
+                self_.action_range = nn.Parameter(
```
Comment (duburcqa) [inline rllib/agents/ddpg/ddpg_torch_model.py:93]: why not use `self.register_buffer` ? since it is not a parameter actually.

File: rllib/agents/ddpg/ddpg_torch_policy.py [hunk 2]
```diff
@@ -21,8 +22,8 @@ def build_ddpg_models_and_action_dist(policy, obs_space, action_space, config):
     # TODO(sven): Unify this once we generically support creating more than
     #  one Model per policy. Note: Device placement is done automatically
     #  already for `policy.model` (but not for the target model).
-    device = (torch.device("cuda")
-              if torch.cuda.is_available() else torch.device("cpu"))
+    device = (torch.device("cuda") if torch.cuda.is_available()
+              and ray.get_gpu_ids() else torch.device("cpu"))
```
Comment (duburcqa) [inline rllib/agents/ddpg/ddpg_torch_policy.py:26]: On my computer, `ray.get_gpu_ids()` is 0, even if the number of gpu is higher than 0. It is non-zero only if a use the decorator `@remote(num_gpus=N)`, with N > 0.

File: rllib/agents/ddpg/ddpg_torch_model.py (hunk 1 above)
Comment (raphaelavalos) [reply rllib/agents/ddpg/ddpg_torch_model.py:93]: Good idea ! I will make the change when I can.

File: rllib/agents/ddpg/ddpg_torch_policy.py (hunk 2 above)
Comment (raphaelavalos) [reply rllib/agents/ddpg/ddpg_torch_policy.py:26]: If I am not wrong, inside a trainer class if you set up the config with the GPU parameter and `ray.init` with gpu > 0, it will return true.

Comment (raphaelavalos) [reply rllib/agents/ddpg/ddpg_torch_policy.py:26]: Also that's how rllib decides to move the main model to the GPU.

Comment (sven1977) [reply rllib/agents/ddpg/ddpg_torch_policy.py:26]: Yes, you have to initialize ray with num_gpus > 0.

**Review by sven1977 — COMMENTED** [review]
Did you add the torch.clamp fix (use torch.min/max) for the case that aciton_space.low/high have different values for different dims?

**Comment by raphaelavalos** [comment]
I forgot to add this. I will make the changes on Monday morning :)

**Comment by raphaelavalos** [comment]
@sven1977 @duburcqa I think I made all the changes we discussed. If you can have a look :)

**Review by duburcqa — APPROVED** [review]

File: rllib/agents/ddpg/ddpg_torch_policy.py [hunk 3]
```diff
@@ -63,11 +64,20 @@ def ddpg_actor_critic_loss(policy, model, _, train_batch):
         clipped_normal_sample = torch.clamp(
             torch.normal(
                 mean=torch.zeros(policy_tp1.size()),
-                std=policy.config["target_noise"]), -target_noise_clip,
-            target_noise_clip)
-        policy_tp1_smoothed = torch.clamp(policy_tp1 + clipped_normal_sample,
-                                          policy.action_space.low.item(0),
-                                          policy.action_space.high.item(0))
+                std=policy.config["target_noise"]).to(policy_tp1.device),
+            -target_noise_clip, target_noise_clip)
```
Comment (duburcqa) [inline rllib/agents/ddpg/ddpg_torch_policy.py:68]: `to(policy_tp1.device)` is still missing

Comment (raphaelavalos) [reply rllib/agents/ddpg/ddpg_torch_policy.py:68]: Clamp returns a tensor on the same device, so it's already on the correct device.

Comment (duburcqa) [reply rllib/agents/ddpg/ddpg_torch_policy.py:68]: Yes, I was talking about the output of `torch.normal`. Maybe I missed it when I wrote this comment. Anyway it is fine in the current commit.

**Comment by raphaelavalos** [comment]
Hello @sven1977 could you check and maybe merge this PR? Thanks

**Comment by duburcqa** [comment]
Too bad this PR is not part of RLlib 0.8.7 :(

**Comment by duburcqa** [comment]
Beside, there is still an issue with GPU support [here](https://github.com/ray-project/ray/blob/78ed5e71e29f6ed479cd143bd3d800465a13957e/rllib/agents/ddpg/ddpg_torch_policy.py#L261). It should be instead:
```
                var_target.data = tau * var.data.to(var_target.device) + \
                    (1.0 - tau) * var_target.data
```

**Comment by raphaelavalos** [comment]
@duburcqa That's weird, both model are on the same device. The device of the target is changed here:

https://github.com/ray-project/ray/blob/78ed5e71e29f6ed479cd143bd3d800465a13957e/rllib/agents/ddpg/ddpg_torch_policy.py#L21-L29

Maybe you can share your code so I can reproduce the issue?

**Comment by duburcqa** [comment]
Ok for the target network, but what about `policy.model` ? It is mentioned that "device placement is done automatically", but I was unable to find where this is done in the code.

**Comment by raphaelavalos** [comment]
It is common to all torch policies:
 
https://github.com/ray-project/ray/blob/edd783bc327760a4892ab89222ee551e42df15b9/rllib/policy/torch_policy.py#L101-L105

**Comment by duburcqa** [comment]
OK I'm using RLlib 0.8.7, so maybe it is the problem. `build_ddpg_models_and_action_dist` fix is not part of this release.

**Comment by raphaelavalos** [comment]
@sven1977 I think the PR is ready to be merged :)

**Comment by raphaelavalos** [comment]
@ericl could you take a look and merge this PR please? :)

**Review by sven1977 — CHANGES_REQUESTED** [review]
Looks good. Just the LINT tests and test_parameter_noise before we can merge this:
https://travis-ci.com/github/ray-project/ray/builds/177577596
